sklearn/metrics/pairwise.py

```diff
@@ -436,6 +437,84 @@ def nan_euclidean_distances(X, Y=None, *, squared=False,
     return distances
 
 
+def _nan_fill_row_norm(r, fill_value: float = 0.):
+    p = r * r
+    p = np.nan_to_num(p, nan=fill_value)
+    return np.sum(p)
+
+
+@njit
+def _nan_fill_dot(X, Y, fill_values: np.ndarray):
+    if X.shape[1] != fill_values.shape[0]:
+        raise ValueError('X and fill_values have incompatible shapes')
+
+    p = np.zeros((X.shape[0], Y.shape[0],))
+    for i in range(0, X.shape[0]):
+        v1 = X[i, :]
+        for j in range(0, Y.shape[0]):
+            v2 = Y[j, :]
+            s = 0.
+            for k in range(0, fill_values.shape[0]):
+                if np.isnan(v1[k]) and np.isnan(v2[k]):
+                    pass
+                elif np.isnan(v1[k]) or np.isnan(v2[k]):
+                    s += fill_values[k]
+                else:
+                    s += v1[k] * v2[k]
+
+            p[i, j] = s
+
+    return p
+
+
+def nan_filled_euclidean_distances(X, fill_values, Y=None, squared=False, missing_values=np.nan, copy=True):
+    if X.dtype == np.float32 or (Y is not None and Y.dtype == np.float32):
```

Comment:
why float32 does not work?

Reply:
It actually does work with float32. I think that was leftover from an earlier version which didn't allow it. I'll remove the check.